### PR TITLE
#139 update changes

### PR DESCRIPTION
**Pull request by llbartekll**
- Adds ControllerSessionStateMachine and NonControllerSessionStateMachine for handling method updates
- flattens session object and proposal and settle request
- remove upgrade method
- adds wc_sessionUpdateAccounts and wc_sessionUpdateMethods

### CODE REVIEW

File: Sources/WalletConnect/Engine/SessionEngine.swift [hunk 1]
```diff
@@ -9,11 +9,10 @@ final class SessionEngine {
     var onSessionResponse: ((Response)->())?
     var onSessionSettle: ((Session)->())?
     var onSessionRejected: ((String, SessionType.Reason)->())?
-    var onSessionUpdate: ((String, Set<Account>)->())?
-    var onSessionUpgrade: ((String, SessionPermissions)->())?
+    var onSessionUpdateAccounts: ((String, Set<Account>)->())?
     var onSessionExtended: ((Session) -> ())?
     var onSessionDelete: ((String, SessionType.Reason)->())?
-    var onNotificationReceived: ((String, Session.Notification)->())?
+    var onNotificationReceived: ((String, Session.Event)->())?
```
Comment (MisterVants) [inline Sources/WalletConnect/Engine/SessionEngine.swift:15]: I think we can change this var to `onEventReceived`, now that we're dropping the "notification" terminology

File: Sources/WalletConnect/Engine/SessionEngine.swift [hunk 2]
```diff
@@ -370,29 +325,28 @@ final class SessionEngine {
         relayer.respondSuccess(for: payload)
     }
     
-    private func wcSessionNotification(_ payload: WCRequestSubscriptionPayload, notificationParams: SessionType.NotificationParams) {
+    private func wcSessionNotification(_ payload: WCRequestSubscriptionPayload, notificationParams: SessionType.EventParams) {
```
Comment (MisterVants) [inline Sources/WalletConnect/Engine/SessionEngine.swift:328]: We can also change this method name to `wcSessionEvent`

File: Sources/WalletConnect/Engine/SessionEngine.swift [hunk 3]
```diff
@@ -446,50 +398,37 @@ final class SessionEngine {
...
-            onSessionUpgrade?(session.topic, permissions)
-        case .error:
-            logger.error("Peer failed to upgrade permissions.")
-        }
-    }
-    
-    private func validatePermissions(_ permissions: SessionPermissions) -> Bool {
-//        for chainId in permissions.blockchain.chains {
-//            if !String.conformsToCAIP2(chainId) {
+//    private func validatePermissions(_ permissions: SessionPermissions) -> Bool {
```
Comment (MisterVants) [inline Sources/WalletConnect/Engine/SessionEngine.swift:414]: Do you still need this commented section for the methods validation?

File: Sources/WalletConnect/Types/Session/SessionSequence.swift [hunk 4]
```diff
@@ -13,8 +13,12 @@ struct SessionSequence: ExpirableSequence {
     let relay: RelayProtocolOptions
     let controller: AgreementPeer
     let participants: Participants
-    var blockchain: SessionType.Blockchain
-    var permissions: SessionPermissions
+
+//    let `self`: Participant
```
Comment (MisterVants) [inline Sources/WalletConnect/Types/Session/SessionSequence.swift:17]: I think we don't this commented code anymore

**Review by MisterVants — APPROVED** [review]

File: Sources/WalletConnect/Types/Session/SessionSequence.swift (hunk 4 above)
Comment (llbartekll) [reply Sources/WalletConnect/Types/Session/SessionSequence.swift:17]: We are going to flatten participants as well so actually it will become valid

**Pull request merged**